File: samples/arena_demo.lua
-- Sapphire Lua 5.1 Runtime Header

local Arena = {}
Arena.__index = Arena

function Arena.init()
  local self = setmetatable({}, Arena)
  self.objects = {}
  return self
end

function Arena:register(obj)
  if type(obj) == "table" then
    local meta = getmetatable(obj)
    if meta and meta.__arena and meta.__arena ~= self then
      for i, o in ipairs(meta.__arena.objects) do
        if o == obj then
          table.remove(meta.__arena.objects, i)
          break
        end
      end
    end
    table.insert(self.objects, obj)
    if meta then
      meta.__arena = self
    end
  end
  return obj
end

function Arena:destroy()
  for _, obj in ipairs(self.objects) do
    local meta = getmetatable(obj)
    if meta and meta.__shadow then
      for k in pairs(meta.__shadow) do
        meta.__shadow[k] = nil
      end
    end
  end
  self.objects = {}
end

local _DEFAULT_ARENA = Arena.init()

local _clone_helper

local function _create_proto_object(proto, class_tbl)
  local shadow = {}
  local meta = {}
  meta.__shadow = shadow
  meta.__proto = proto
  meta.__class = class_tbl
  meta.__index = function(tbl, key)
    if key == "__proto__" then return meta.__proto end
    if key == "__shadow__" then return meta.__shadow end
    if shadow[key] ~= nil then return shadow[key] end
    if proto ~= nil then
      local val = proto[key]
      if val ~= nil then
        if type(val) == "table" then
          local cow = _clone_helper(val, nil, nil)
          shadow[key] = cow
          return cow
        end
        return val
      end
    end
    if class_tbl and class_tbl[key] ~= nil then
      return class_tbl[key]
    end
    return nil
  end
  meta.__newindex = function(tbl, key, val)
    if key == "__proto__" or key == "__shadow__" then return end
    if proto ~= nil then
      shadow[key] = val
    else
      rawset(tbl, key, val)
    end
  end
  local obj = setmetatable({}, meta)
  if proto == nil then
    _DEFAULT_ARENA:register(obj)
  end
  return obj
end

_clone_helper = function(obj, init_fn, arena)
  if obj == nil then return nil end
  local clone_obj
  if type(obj) == "table" then
    local meta = getmetatable(obj)
    if obj.clone and type(obj.clone) == "function" then
      clone_obj = obj:clone()
    elseif meta and meta.__class then
      clone_obj = _create_proto_object(obj, meta.__class)
    else
      clone_obj = _create_proto_object(obj, nil)
    end
  else
    clone_obj = obj
  end
  if arena ~= nil and type(clone_obj) == "table" then
    arena:register(clone_obj)
  end
  if init_fn then
    init_fn(clone_obj)
  end
  return clone_obj
end


local Point = {}
Point.__index = Point
function Point.init(kwargs, proto)
  kwargs = kwargs or {}
  local self
  self = setmetatable({}, Point)
  if proto == nil then
  end
  for k, v in pairs(kwargs) do
    self[k] = v
  end
  if Point._init_sapphire then
    Point._init_sapphire(self, (kwargs['x'] ~= nil and kwargs['x'] or kwargs[1]), (kwargs['y'] ~= nil and kwargs['y'] or kwargs[2]))
  end
  return self
end

function Point:_init_sapphire(x, y)
  if x == nil then x = 0.0 end
  if y == nil then y = 0.0 end
  self.x = x
  self.y = y
end

function Point:translate(dx, dy)
  self.x = self.x + dx
  self.y = self.y + dy
end


local Particle = {}
Particle.__index = Particle
function Particle.init(kwargs, proto)
  kwargs = kwargs or {}
  local self
  self = setmetatable({}, Particle)
  if proto == nil then
  end
  for k, v in pairs(kwargs) do
    self[k] = v
  end
  if Particle._init_sapphire then
    Particle._init_sapphire(self, (kwargs['pos'] ~= nil and kwargs['pos'] or kwargs[1]), (kwargs['lifetime'] ~= nil and kwargs['lifetime'] or kwargs[2]))
  end
  return self
end

function Particle:_init_sapphire(pos, lifetime)
  if lifetime == nil then lifetime = 1.0 end
  self.pos = pos
  self.lifetime = lifetime
end


local Entity = {}
Entity.__index = Entity
function Entity.init(kwargs, proto)
  kwargs = kwargs or {}
  local self
  self = _create_proto_object(proto, Entity)
  if proto == nil then
  end
  for k, v in pairs(kwargs) do
    self[k] = v
  end
  return self
end


local Enemy = {}
Enemy.__index = Enemy
function Enemy.init(kwargs, proto)
  kwargs = kwargs or {}
  local self
  self = _create_proto_object(proto, Enemy)
  if proto == nil then
  end
  for k, v in pairs(kwargs) do
    self[k] = v
  end
  if Enemy._init_sapphire then
    Enemy._init_sapphire(self, (kwargs['name'] ~= nil and kwargs['name'] or kwargs[1]), (kwargs['hp'] ~= nil and kwargs['hp'] or kwargs[2]), (kwargs['pos'] ~= nil and kwargs['pos'] or kwargs[3]), (kwargs['damage'] ~= nil and kwargs['damage'] or kwargs[4]))
  end
  return self
end

function Enemy:_init_sapphire(name, hp, pos, damage)
  if damage == nil then damage = 10 end
  self.name = name
  self.hp = hp
  self.pos = pos
  self.damage = damage
end

function Enemy:take_damage(amount)
  self.hp = self.hp - amount
  if (self.hp < 0) then
    self.hp = 0
  end
end


local Boss = {}
Boss.__index = Boss
function Boss.init(kwargs, proto)
  kwargs = kwargs or {}
  local self
  self = _create_proto_object(proto, Boss)
  if proto == nil then
  end
  for k, v in pairs(kwargs) do
    self[k] = v
  end
  if Boss._init_sapphire then
    Boss._init_sapphire(self, (kwargs['name'] ~= nil and kwargs['name'] or kwargs[1]), (kwargs['hp'] ~= nil and kwargs['hp'] or kwargs[2]), (kwargs['pos'] ~= nil and kwargs['pos'] or kwargs[3]), (kwargs['phase'] ~= nil and kwargs['phase'] or kwargs[4]))
  end
  return self
end

function Boss:_init_sapphire(name, hp, pos, phase)
  if phase == nil then phase = 1 end
  self.name = name
  self.hp = hp
  self.pos = pos
  self.phase = phase
end


local function demo_explicit_arenas()
  local level_arena = Arena.init()
  local combat_arena = Arena.init()
  local spawn_point = level_arena:register(Point.init({x = 100.0, y = 200.0}))
  local base_goblin = level_arena:register(Enemy.init({name = "Goblin Archer", hp = 50, pos = spawn_point, damage = 12}))
  local minion = _clone_helper(base_goblin, function(self)
    self.hp = 30
  end)
  local combat_goblin = _clone_helper(base_goblin, function(self)
    self.hp = 75
  end, combat_arena)
  minion:take_damage(10)
  combat_goblin:take_damage(25)
  combat_arena:destroy()
  level_arena:destroy()
  return (minion.hp + combat_goblin.hp)
end


local function demo_scoped_raii_cleanup()
  local total_lifetime = 0.0
  local temp_arena = Arena.init()
  local origin = temp_arena:register(Point.init({x = 5.0, y = 10.0}))
  local p1 = temp_arena:register(Particle.init({pos = origin, lifetime = 1.5}))
  local p2 = temp_arena:register(Particle.init({pos = origin, lifetime = 2.5}))
  local temp_boss = temp_arena:register(Boss.init({name = "Dungeon Boss", hp = 500, pos = origin, phase = 1}))
  total_lifetime = (p1.lifetime + p2.lifetime)
  temp_arena:destroy()
  return total_lifetime
end


local function demo_implicit_default_arena()
  local default_pos = Point.init({x = 0.0, y = 0.0})
  local default_enemy = Enemy.init({name = "Default Skeleton", hp = 100, pos = default_pos})
  local cloned_skeleton = _clone_helper(default_enemy, function(self)
    self.hp = 60
  end)
  cloned_skeleton:take_damage(15)
  return (default_enemy.hp + cloned_skeleton.hp)
end


local function run_demo()
  local score1 = demo_explicit_arenas()
  local score2 = demo_implicit_default_arena()
  local duration = demo_scoped_raii_cleanup()
  local result = (score1 + score2)
  if (duration > 3.0) then
    result = result + 10
  end
  return result
end


run_demo()

File: samples/arena_demo.py
# Sapphire Runtime Header
import copy
from enum import IntEnum

class Arena:
  def __init__(self):
    self.objects = []
  def register(self, obj):
    if hasattr(obj, '__arena__') and obj.__arena__ is not None:
      if obj.__arena__ is not self:
        try:
          obj.__arena__.objects.remove(obj)
        except ValueError:
          pass
    if obj not in self.objects:
      self.objects.append(obj)
    if hasattr(obj, '__setattr__'):
      try:
        object.__setattr__(obj, '__arena__', self)
      except AttributeError:
        pass
    return obj
  def destroy(self):
    for obj in self.objects:
      if hasattr(obj, '__shadow__'):
        obj.__shadow__.clear()
    self.objects.clear()
  def __enter__(self):
    return self
  def __exit__(self, exc_type, exc_val, exc_tb):
    self.destroy()

_DEFAULT_ARENA = Arena()

class SapphireObject:
  def __init__(self, proto=None):
    super().__setattr__('__proto__', proto)
    super().__setattr__('__shadow__', {})
    if proto is None:
      _DEFAULT_ARENA.register(self)

  def clone(self):
    clone_obj = self.__class__(proto=self)
    if hasattr(self, '__arena__') and self.__arena__ is not None:
      self.__arena__.register(clone_obj)
    return clone_obj

  def __getattr__(self, name):
    if name.startswith('__') and name.endswith('__'):
      if name == '__proto__':
        return self.__proto__
      raise AttributeError(f"Attribute '{name}' not found on {self.__class__.__name__}")
    if name in self.__shadow__:
      return self.__shadow__[name]
    if self.__proto__ is not None:
      val = getattr(self.__proto__, name)
      if not isinstance(val, (int, float, bool, str, type(None))):
        if hasattr(val, 'clone'):
          cow_val = val.clone()
        else:
          cow_val = copy.deepcopy(val)
        self.__shadow__[name] = cow_val
        return cow_val
      return val
    raise AttributeError(f"Attribute '{name}' not found on {self.__class__.__name__}")

  def __setattr__(self, name, value):
    if name in ('__proto__', '__shadow__'):
      super().__setattr__(name, value)
    elif self.__proto__ is not None:
      self.__shadow__[name] = value
    else:
      super().__setattr__(name, value)

def _clone_helper(obj, init_fn=None, arena=None):
  if arena is None and hasattr(obj, '__arena__'):
    arena = getattr(obj, '__arena__', None)
  clone_obj = obj.clone()
  if arena is not None:
    arena.register(clone_obj)
  if init_fn:
    init_fn(clone_obj)
  return clone_obj


class Point(object):
  def __init__(self, *args, **kwargs):
    for k, v in kwargs.items():
      setattr(self, k, v)
    self._init_sapphire(*args, **kwargs)
  def _init_sapphire(self, x=0.0, y=0.0):
    self.x = x
    self.y = y
  def translate(self, dx, dy):
    self.x += dx
    self.y += dy


class Particle(object):
  def __init__(self, *args, **kwargs):
    for k, v in kwargs.items():
      setattr(self, k, v)
    self._init_sapphire(*args, **kwargs)
  def _init_sapphire(self, pos, lifetime=1.0):
    self.pos = pos
    self.lifetime = lifetime


class Entity(SapphireObject):
  def __init__(self, *args, proto=None, **kwargs):
    super().__init__(proto=proto)
    if proto is None:
      for k, v in kwargs.items():
        setattr(self, k, v)
  pass


class Enemy(Entity):
  def __init__(self, *args, proto=None, **kwargs):
    super().__init__(proto=proto)
    if proto is None:
      for k, v in kwargs.items():
        setattr(self, k, v)
      self._init_sapphire(*args, **kwargs)
  def _init_sapphire(self, name, hp, pos, damage=10):
    self.name = name
    self.hp = hp
    self.pos = pos
    self.damage = damage
  def take_damage(self, amount):
    self.hp -= amount
    if (self.hp < 0):
      self.hp = 0


class Boss(Entity):
  def __init__(self, *args, proto=None, **kwargs):
    super().__init__(proto=proto)
    if proto is None:
      for k, v in kwargs.items():
        setattr(self, k, v)
      self._init_sapphire(*args, **kwargs)
  def _init_sapphire(self, name, hp, pos, phase=1):
    self.name = name
    self.hp = hp
    self.pos = pos
    self.phase = phase


def demo_explicit_arenas():
  level_arena = Arena()
  try:
    combat_arena = Arena()
    try:
      spawn_point = level_arena.register(Point(x=100.0, y=200.0))
      base_goblin = level_arena.register(Enemy(name="Goblin Archer", hp=50, pos=spawn_point, damage=12))
      minion = _clone_helper(base_goblin, lambda self: [setattr(self, 'hp', 30)])
      combat_goblin = _clone_helper(base_goblin, lambda self: [setattr(self, 'hp', 75)], arena=combat_arena)
      minion.take_damage(10)
      combat_goblin.take_damage(25)
      return (minion.hp + combat_goblin.hp)
    finally:
      combat_arena.destroy()
  finally:
    level_arena.destroy()

def demo_scoped_raii_cleanup():
  total_lifetime = 0.0
  temp_arena = Arena()
  try:
    origin = temp_arena.register(Point(x=5.0, y=10.0))
    p1 = temp_arena.register(Particle(pos=origin, lifetime=1.5))
    p2 = temp_arena.register(Particle(pos=origin, lifetime=2.5))
    temp_boss = temp_arena.register(Boss(name="Dungeon Boss", hp=500, pos=origin, phase=1))
    total_lifetime = (p1.lifetime + p2.lifetime)
  finally:
    temp_arena.destroy()
  return total_lifetime

def demo_implicit_default_arena():
  default_pos = Point(x=0.0, y=0.0)
  default_enemy = Enemy(name="Default Skeleton", hp=100, pos=default_pos)
  cloned_skeleton = _clone_helper(default_enemy, lambda self: [setattr(self, 'hp', 60)])
  cloned_skeleton.take_damage(15)
  return (default_enemy.hp + cloned_skeleton.hp)

def run_demo():
  score1 = demo_explicit_arenas()
  score2 = demo_implicit_default_arena()
  duration = demo_scoped_raii_cleanup()
  result = (score1 + score2)
  if (duration > 3.0):
    result += 10
  return result

if __name__ == "__main__":
  
  run_demo()
